pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1374,13 +1388,13 @@ func (c *PodAffinityChecker) satisfiesPodsAffinityAntiAffinity(pod *v1.Pod,
 		// Check all affinity terms.
 		topologyPairsPotentialAffinityPods := predicateMeta.topologyPairsPotentialAffinityPods
 		if affinityTerms := GetPodAffinityTerms(affinity.PodAffinity); len(affinityTerms) > 0 {
-			matchExists := c.nodeMatchesAllTopologyTerms(pod, topologyPairsPotentialAffinityPods, nodeInfo, affinityTerms)
+			matchExists := c.nodeMatchesAllTopologyTerms(topologyPairsPotentialAffinityPods, nodeInfo, affinityTerms)
 			if !matchExists {
 				// This pod may the first pod in a series that have affinity to themselves. In order
-				// to not leave such pods in pending state forever, we check that if no other pod
+				// to not leave such pods in pending state forever, we check that if no other node
 				// in the cluster matches the namespace and selector of this pod and the pod matches
 				// its own terms, then we allow the pod to pass the affinity check.
-				if !(len(topologyPairsPotentialAffinityPods.topologyPairToPods) == 0 && targetPodMatchesAffinityOfPod(pod, pod)) {
+				if predicateMeta.isPodAffinitySatisfied || !podMatchesItsOwnAffinityOnNode(pod, node) {
```

Comment:
seems moving this check to line 1388 makes more sense. If so the logic'd be:

- if incomingPod's podAffinity is satisfied, but we have no idea which nodes are fits, so continue to see if current node is a fit
- if incomingPod's podAffinity is not satisfied, then for sure it's not a fit on current node, so don't need to calculate `matchExists`; instead, directly check if `podMatchesItsOwnAffinityOnNode`

cc @bsalamat for brainstorming